src/sync.rs

```diff
@@ -352,36 +352,32 @@ where
     optimize(schedule, &mut sync);
 
     let mut result = Schedule::default();
-    let mut signals: HashMap<Semaphore, Option<S>> = HashMap::new();
-    let mut waits: HashMap<Semaphore, Option<W>> = HashMap::new();
+    let mut signals: HashMap<Semaphore, S> = HashMap::new();
```

Comment:
Why this one was an issue?
Assertions were here to guarantee that you never create signal/wait pair for the same `Semaphore`. They are expected to be unique.

Reply:
It used to `take` from the `HashMap`s, which does not remove the actual entries in the `HashMap`, just setting the value to `None`. `remove` returns the value, and leaves the map empty at the end, so, it's exactly what this should do.